tools/ops/policystream.py

```diff
@@ -0,0 +1,640 @@
+        """
+        if target == 'HEAD':
+            target = self.repo.head.target
+
+        commits = []
+        for commit in self.repo.walk(target, sort):
+            log.debug(
+                "processing commit id:%s date:%s parents:%d msg:%s",
+                str(commit.id)[:6], commit_date(commit).isoformat(),
+                len(commit.parents), commit.message)
+            commits.append(commit)
+            if limit and len(commits) > limit:
+                break
+
+        if limit:
+            self.initialize_tree(commits[limit].tree)
+            commits.pop(-1)
+
+        for commit in commits:
+            for policy_change in self.process_commit(commit):
+                yield policy_change
+
+    def process_commit(self, change):
+        if not change.parents:
+            change_diff = self.repo.diff(self.repo.get(EMPTY_TREE, change), change)
+        else:
+            change_diff = self.repo.diff(change.parents[0], change)
+
+        log.debug(
+            "processing commit id:%s date:%s parents:%d add:%d del:%d files:%d change:%s",
+            str(change.id)[:6], commit_date(change).isoformat(), len(change.parents),
+            change_diff.stats.insertions,
+            change_diff.stats.deletions,
+            change_diff.stats.files_changed,
+            change.message.strip())
+
+        change_policies = PolicyCollection()
+        current_policies = PolicyCollection()
+        removed = set()
+
+        for delta in change_diff.deltas:
+            f = delta.new_file.path
+            if not self.matcher(f):
+                continue
+            elif delta.status == GIT_DELTA_INVERT['GIT_DELTA_ADDED']:
+                change_policies += PolicyCollection.from_data(
+                    yaml.safe_load(self.repo.get(change.tree[f].id).data),
+                    Config.empty(), f)
+            elif delta.status == GIT_DELTA_INVERT['GIT_DELTA_MODIFIED']:
+                change_policies += PolicyCollection.from_data(
+                    yaml.safe_load(self.repo.get(change.tree[f].id).data),
+                    Config.empty(), f)
+                current_policies += self.policy_files[f]
+            elif delta.status == GIT_DELTA_INVERT['GIT_DELTA_DELETED']:
+                current_policies += self.policy_files[f]
+                removed.add(f)
+            elif delta.status == GIT_DELTA_INVERT['GIT_DELTA_RENAMED']:
+                change_policies += PolicyCollection.from_data(
+                    yaml.safe_load(self.repo.get(change.tree[f].id).data),
+                    Config.empty(), f)
+                current_policies += self.policy_files[delta.old_file.path]
+                removed.add(delta.old_file.path)
+            else:
+                log.info(
+                    "unhandled delta type:%s path:%s commit_id:%s",
+                    GIT_DELTA[delta.status], delta.new_file.path, change.id)
+                continue
+
+        for pchange in CollectionDelta(
```

Comment:
this could be separated out into a separate method, _process_steam_change as its just updating internal bookeeping data structs.